pkg/controller/deployment/deployment_controller.go

```diff
@@ -564,6 +567,58 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+// proportionalScaling should run only on scaling events and not during the normal rollout process.
+func (dc *DeploymentController) proportionalScaling(deployment extensions.Deployment) error {
+	newRS, oldRSs, err := dc.getAllReplicaSets(deployment)
+	if err != nil {
+		return err
+	}
+	if len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+		// No old replica sets with pods; scale new replica set to the full count.
+		_, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment)
+		return err
+	}
+	allRSs := append(controller.FilterActiveReplicaSets(oldRSs), newRS)
+
+	sort.Sort(controller.ByAvailableReplicas(allRSs))
+
+	maybeMaxSurge := 0
+	if deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType {
+		maybeMaxSurge, err = intstrutil.GetValueFromIntOrPercent(&deployment.Spec.Strategy.RollingUpdate.MaxSurge, deployment.Spec.Replicas)
```

Comment:
This will be wrong though if newRs.Spec.Replicas is lower than d.Spec.Replicas. For example:
```
maxSurge is 1, d.Spec.Replicas is 10, newRs.Spec.Replicas is 6
new maxSurge value would be max(min(1, 6-10), 0)=>max(min(1,-4), 0)=>max(-4, 0)=>0
```
I think you meant `IntMax(IntMin(maybeMaxSurge, d.Spec.Replicas-newRs.Spec.Replicas), 0)`:
```
maxSurge is 1, d.Spec.Replicas is 10, newRs.Spec.Replicas is 6
new maxSurge value would be max(min(1, 10-6), 0)=>max(min(1,4), 0)=>max(1, 0)=>1

maxSurge is 1, d.Spec.Replicas is 10, newRs.Spec.Replicas is 10
new maxSurge value would be max(min(1, 10-10), 0)=>max(min(1,0), 0)=>max(0, 0)=>0

maxSurge is 1, d.Spec.Replicas is 10, newRs.Spec.Replicas is 12
new maxSurge value would be max(min(1, 10-12), 0)=>max(min(1,-2), 0)=>max(-2, 0)=>0
```